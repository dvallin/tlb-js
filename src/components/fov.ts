import { SetSpace } from '../spatial/set-space'

export interface FovComponent {
  fov: SetSpace
}
