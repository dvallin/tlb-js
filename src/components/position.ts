import { Vector } from '../spatial'

export interface PositionComponent {
  level: number
  position: Vector
}

export function centeredPosition(level: number, position: Vector): PositionComponent {
  return { level, position: position.center }
}
