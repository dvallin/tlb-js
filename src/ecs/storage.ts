
export interface Storage<T> {

    insert(entity: number, value: T): void
    get(entity: number): T | undefined
    remove(entity: number): T | undefined
    has(entity: number): boolean

    foreach(f: (entity: number, value: T) => void): void
    clear(): void
}

export class SetStorage implements Storage<{}> {

    private data: Set<number> = new Set()

    public insert(entity: number, { }: {}): void {
        this.data.add(entity)
    }

    public clear(): void {
        this.data.clear()
    }

    public get(entity: number): {} | undefined {
        return this.data.has(entity) ? {} : undefined
    }

    public remove(entity: number): {} | undefined {
        return this.data.delete(entity) ? {} : undefined
    }

    public has(entity: number): boolean {
        return this.data.has(entity)
    }

    public foreach(f: (entity: number, value: {}) => void): void {
        this.data.forEach(v => f(v, {}))
    }
}

export class MapStorage<T> implements Storage<T> {

    private data: Map<number, T> = new Map()

    public insert(entity: number, value: T): void {
        this.data.set(entity, value)
    }

    public clear(): void {
        this.data.clear()
    }

    public get(entity: number): T | undefined {
        return this.data.get(entity)
    }

    public remove(entity: number): T | undefined {
        const value = this.get(entity)
        this.data.delete(entity)
        return value
    }

    public has(entity: number): boolean {
        return this.data.has(entity)
    }

    public foreach(f: (entity: number, value: T) => void): void {
        this.data.forEach((value, entity) => f(entity, value))
    }
}

export class VectorStorage<T> implements Storage<T> {

    private data: (T | undefined)[] = []

    public insert(entity: number, value: T): void {
        this.data[entity] = value
    }

    public clear(): void {
        this.data = []
    }

    public get(entity: number): T | undefined {
        return this.data[entity]
    }

    public remove(entity: number): T | undefined {
        const value = this.get(entity)
        this.data[entity] = undefined
        return value
    }

    public has(entity: number): boolean {
        return this.data[entity] !== undefined
    }

    public foreach(f: (entity: number, value: T) => void): void {
        for (let entity = 0; entity < this.data.length; entity++) {
            const value = this.data[entity]
            if (value !== undefined) {
                f(entity, value)
            }
        }
    }
}
