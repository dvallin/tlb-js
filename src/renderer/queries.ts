import { FOV, Lighting } from 'rot-js'
import { TlbWorld } from '../tlb'
import { WorldMap, WorldMapResource } from '../resources/world-map'
import { Vector } from '../spatial'
import { Color } from './color'
import { bfs } from './bfs'
import { FunctionalShape } from '../geometry/functional-shape'
import { astar, Path } from './astar'
import { LineSegment } from '../geometry/line-segment'

export interface QueryParameters {
  onlyDiscovered: boolean
  maximumCost: number
  bestEffort: boolean
}

export class Queries {
  public fov(world: TlbWorld, origin: Vector, callback: (pos: Vector, distance: number) => void) {
    const originFloor = origin.floor()
    const map: WorldMap = world.getResource<WorldMapResource>('map')
    const fov = new FOV.RecursiveShadowcasting((x, y) => !map.isLightBlocking(world, new Vector(x, y)), { topology: 8 })
    const seenAlready = new Set<string>()
    fov.compute(originFloor.x, originFloor.y, 20, (x, y, distance) => {
      const position = new Vector(x, y)
      const key = position.key
      if (!seenAlready.has(key)) {
        seenAlready.add(key)
        callback(position, distance)
      }
    })
  }

  public lighting(world: TlbWorld, origin: Vector, color: Color, callback: (pos: Vector, color: Color) => void) {
    const originFloor = origin.floor()
    const map: WorldMap = world.getResource<WorldMapResource>('map')
    const fov = new FOV.RecursiveShadowcasting((x, y) => !map.isLightBlocking(world, new Vector(x, y)), { topology: 8 })
    const lighting = new Lighting((x, y) => (map.isLightBlocking(world, new Vector(x, y)) ? 0.0 : 1.0), { passes: 1 })
    lighting.setLight(originFloor.x, originFloor.y, color.color)
    lighting.setFOV(fov)
    lighting.setOptions({ range: 6 })
    lighting.compute((x: number, y: number, c: [number, number, number]) => {
      callback(new Vector(x, y), new Color(c))
    })
  }

  public explore(
    world: TlbWorld,
    origin: Vector,
    visit: (pos: Vector, distance: number) => boolean,
    params: Partial<QueryParameters>
  ): void {
    const originFloor = origin.floor()
    const map: WorldMap = world.getResource<WorldMapResource>('map')
    const maximumDepth = params.maximumCost || Number.MAX_SAFE_INTEGER
    const onlyDiscovered = params.onlyDiscovered || false
    bfs(originFloor, target => FunctionalShape.lN(target, 1, false), visit, (target, depth) => {
      if (depth >= maximumDepth || map.isBlocking(world, target)) {
        return false
      }
      return onlyDiscovered ? map.isDiscovered(target) : true
    })
  }

  public shortestPath(world: TlbWorld, origin: Vector, target: Vector, params: Partial<QueryParameters>): Path | undefined {
    const map: WorldMap = world.getResource<WorldMapResource>('map')
    const onlyDiscovered = params.onlyDiscovered || false
    const originFloor = origin.floor()
    const targetFloor = target.floor()
    const maximumCost = params.maximumCost || Number.MAX_SAFE_INTEGER
    const bestEffort = params.bestEffort || false
    const path = astar(
      originFloor,
      targetFloor,
      (current, neighbour) => {
        if (map.isBlocking(world, neighbour)) {
          return undefined
        }
        if (onlyDiscovered && !map.isDiscovered(neighbour)) {
          return undefined
        }
        return current.minus(neighbour).l1()
      },
      v => FunctionalShape.lN(v, 1, false),
      v => targetFloor.minus(v).lN(),
      maximumCost,
      bestEffort
    )
    if (path !== undefined && path.path.length - 1 <= maximumCost) {
      return { path: path.path.slice(0, path.path.length - 1), cost: path.cost }
    }
    return undefined
  }

  public ray(world: TlbWorld, origin: Vector, target: Vector, params: Partial<QueryParameters>): Path | undefined {
    const map: WorldMap = world.getResource<WorldMapResource>('map')
    const originFloor = origin.floor()
    const targetFloor = target.floor()
    const maximumCost = params.maximumCost || Number.MAX_SAFE_INTEGER
    const path: Vector[] = []
    const success = new LineSegment(targetFloor, originFloor).all(p => {
      path.push(p)
      return !map.isLightBlocking(world, p, false)
    })
    if (success) {
      let cost = path.length - 1
      if (path.length - 1 <= maximumCost) {
        return { path: path.slice(0, path.length - 1), cost: cost }
      }
    }
    return undefined
  }
}
