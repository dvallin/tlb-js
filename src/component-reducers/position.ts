import { WorldMapResource } from '../resources/world-map'
import { Vector } from '../spatial'
import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { PositionComponent } from '../components/position'

export function move(world: TlbWorld, entity: Entity, level: number, position: Vector): void {
  const l = world.getResource<WorldMapResource>('map').levels[level]
  if (l !== undefined && !l.isBlocking(world, position, entity)) {
    const oldPosition = world.getComponent<PositionComponent>(entity, 'position')!
    l.removeCharacter(oldPosition.position)
    l.setCharacter(position, entity)
    world.editEntity(entity).withComponent<PositionComponent>('position', { level, position })
  }
}
