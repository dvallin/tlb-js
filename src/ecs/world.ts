import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
