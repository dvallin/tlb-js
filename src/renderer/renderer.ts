import { Display } from 'rot-js'

import { gray } from './palettes'
import { Position } from './position'
import { Color } from './color'

import { Viewport, ViewportResource } from '../resources/viewport'
import { TlbWorld } from '../tlb'
import { getFeature, Feature } from '../components/feature'
import { PositionComponent } from '../components/position'
import { Entity } from '../ecs/entity'
import { WorldMap, WorldMapResource } from '../resources/world-map'
import { LightingComponent } from '../components/light'
import { OverlayComponent } from '../components/overlay'
import { UI, UIResource } from '../resources/ui'

export interface Renderer {
  render(world: TlbWorld): void

  clear(): void
  eventToPosition(e: UIEvent): Position | undefined

  character(character: string, position: Position, fg: Color, bg?: Color): void
  text(text: string, position: Position, fg: Color, bg?: Color): void
}

export class RotRenderer implements Renderer {
  public readonly display: Display

  public ambientColor: Color

  public constructor() {
    const displayOptions = {
      width: 60,
      height: 40,
      forceSquareRatio: true,
      fontSize: 17,
      fontFamily: 'Lucida Console, Monaco, monospace',
      bg: gray[4].rgb,
    }
    this.ambientColor = new Color([120, 120, 120])
    this.display = new Display(displayOptions)
    document.body.appendChild(this.display.getContainer() as Node)
  }

  public clear(): void {
    this.display.clear()
  }

  public render(world: TlbWorld): void {
    this.clear()
    const viewport: Viewport = world.getResource<ViewportResource>('viewport')
    viewport.collectRenderables(world).forEach(({ entity, centered }) => {
      if (entity !== undefined) {
        this.renderEntity(world, viewport, entity, centered)
      }
    })
    const ui: UI = world.getResource<UIResource>('ui')
    ui.render(this)
    world.components.get('lighting')!.clear()
    world.components.get('overlay')!.clear()
  }

  public renderEntity(world: TlbWorld, viewport: Viewport, entity: Entity, centered: boolean): void {
    const feature = getFeature(world, entity)
    const position = world.getComponent<PositionComponent>(entity, 'position')
    if (feature && position) {
      this.renderFeature(world, viewport, entity, centered, feature, position)
    }
  }

  public renderFeature(
    world: TlbWorld,
    viewport: Viewport,
    entity: Entity,
    centered: boolean,
    feature: Feature,
    position: PositionComponent
  ): void {
    const map: WorldMap = world.getResource<WorldMapResource>('map')
    const p = position.position.floor()
    if (map.isDiscovered(p)) {
      let lighting = undefined
      if (map.isVisible(p)) {
        lighting = world.getComponent<LightingComponent>(entity, 'lighting')
      }
      const overlay = world.getComponent<OverlayComponent>(entity, 'overlay') || { background: undefined }
      const displayPosition = viewport.toDisplay(position.position, centered)
      this.character(
        feature.character,
        displayPosition,
        this.computeColor(this.ambientColor, feature.diffuse, lighting),
        overlay.background
      )
    }
  }

  public computeColor(ambientLight: Color, diffuse: Color, lighting: LightingComponent | undefined): Color {
    if (lighting === undefined) {
      return diffuse.multiply(ambientLight)
    } else {
      let totalLight = ambientLight
      lighting.incomingLight.forEach(incoming => {
        totalLight = totalLight.add(incoming)
      })
      return diffuse.multiply(totalLight)
    }
  }

  public eventToPosition(e: TouchEvent | MouseEvent): Position | undefined {
    const p = this.display.eventToPosition(e) as [number, number]
    if (typeof p === 'object') {
      return { x: p[0], y: p[1] }
    }
    return undefined
  }

  public character(character: string, position: Position, fg: Color, bg?: Color): void {
    const fgRgb = fg.rgb
    const bgRgb = bg ? bg.rgb : undefined
    this.display.draw(position.x, position.y, character[0], fgRgb, bgRgb || null)
  }

  public text(text: string, position: Position, fg: Color, bg?: Color): void {
    const fgRgb = fg.rgb
    const bgRgb = bg ? bg.rgb : undefined
    for (let idx = 0; idx < text.length; idx++) {
      this.display.draw(position.x + idx, position.y, text[idx], fgRgb, bgRgb || null)
    }
  }
}
