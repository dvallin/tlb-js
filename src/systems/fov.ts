import { ComponentName, TlbSystem, TlbWorld } from '../tlb'
import { PositionComponent } from '../components/position'
import { FovComponent } from '../components/fov'
import { Queries } from '../renderer/queries'
import { Entity } from '../ecs/entity'

export class Fov implements TlbSystem {
  public readonly components: ComponentName[] = ['fov', 'position']

  public constructor(public readonly queries: Queries) {}

  public update(world: TlbWorld, entity: Entity): void {
    const fov = world.getComponent<FovComponent>(entity, 'fov')!
    const position = world.getComponent<PositionComponent>(entity, 'position')!
    fov.fov = []
    this.queries.fov(world, position.position, (position, distance) => fov.fov.push({ position, distance }))
  }
}
