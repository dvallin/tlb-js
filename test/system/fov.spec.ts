import { Fov } from '../../src/systems/fov'
import { TlbWorld } from '../../src/tlb'
import { World } from '../../src/ecs/world'
import { mockComponent, mockReturnValue, mockImplementation3, mockQueries } from '../mocks'
import { FovComponent } from '../../src/components/fov'
import { Storage } from '../../src/ecs/storage'
import { PositionComponent } from '../../src/components/position'
import { Vector } from '../../src/spatial'
import { Queries } from '../../src/renderer/queries'

describe('Fov', () => {
  let world: TlbWorld

  let queries: Queries
  let fov: FovComponent
  let position: PositionComponent
  beforeEach(() => {
    world = new World()

    const positions: Storage<PositionComponent> = mockComponent<PositionComponent>(world, 'position')
    const fovs: Storage<FovComponent> = mockComponent<FovComponent>(world, 'fov')

    position = { position: new Vector([1.2, 1.2]) }
    fov = { fov: [] }

    mockReturnValue<PositionComponent>(positions.get, position)
    mockReturnValue<FovComponent>(fovs.get, fov)

    queries = mockQueries()
    mockImplementation3(queries.fov, ({}, {}, callback: (pos: Vector, distance: number) => void) => {
      callback(new Vector([1, 1]), 2)
      callback(new Vector([0, 0]), 1)
    })

    new Fov(queries).update(world, 0)
  })

  it('pushes all fov cells', () => {
    expect(fov.fov).toEqual([{ position: new Vector([1, 1]), distance: 2 }, { position: new Vector([0, 0]), distance: 1 }])
  })
})
