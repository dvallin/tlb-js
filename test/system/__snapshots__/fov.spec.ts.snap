// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Npc pushes fov 1`] = `
Uint8Array [
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  0,
  0,
  0,
  0,
  0,
]
`;
